Source/IGListCollectionViewLayout.h

```diff
@@ -0,0 +1,94 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import <UIKit/UIKit.h>
+
+#import <IGListKit/IGListMacros.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ This UICollectionViewLayout subclass is for vertically-scrolling lists of data with variable widths and heights. It
+ supports an infinite number of sections and items. All work is done on the main thread, and while extremely efficient,
+ care must be taken not to stall the main thread in sizing delegate methods.
+
+ This layout piggybacks on the mechanics of UICollectionViewFlowLayout in that:
+
+ - Your UICollectionView data source must also conform to UICollectionViewDelegateFlowLayout
+ - Header support given via UICollectionElementKindSectionHeader
+
+ All UICollectionViewDelegateFlowLayout methods are required and used by this layout:
+
+ - (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath;
+ - (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section;
+ - (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section;
+ - (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section;
+ - (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section;
+
+ Sections and items are put into the same horizontal row until the max-y position of an item extends beyond the width
+ of the collection view. When that happens, the item is "newlined" to the next row. The y position of that row is
+ deteremined by the maximum height (including section insets) of the section/item of the previous row.
+
+ Ex. of a section (2,0) with a large width causing a newline.
+ |[ 0,0 ][ 1,0 ]         |
+ |[         2,0         ]|
+
+ A section with a non-zero height header will always cause that section to newline. Headers are always stretched to the
+ width of the collection view, pinched with the section insets.
+
+ Ex. of a section (2,0) with a header inset on the left/right.
+ |[ 0,0 ][ 1,0 ]         |
+ | >======header=======< |
+ | [ 2,0 ]               |
+
+ Section insets apply to items in the section no matter if they begin on a new row or are on the same row as a previous
+ section.
+
+ Ex. of a section (2) with multiple items and a left inset.
+ |[ 0,0 ][ 1,0 ] >[ 2,0 ]|
+ | >[ 2,1 ][ 2,2 ][ 2,3 ]|
+
+ Interitem spacing applies to items and sections within the same row. Line spacing only applies to items within the same
+ section.
+
+ Please see the unit tests for more configuration examples and expected output.
+ */
+IGLK_SUBCLASSING_RESTRICTED
```

Comment:
Hmm, maybe we could. It's not really designed to be subclassed, but I guess we could let people do that if they wanted. @jessesquires any thoughts?